OKTA-809423 - Create system log query guide (#5099)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/system-log-query/index.md

### User events

| Event                       | Description                                               |
| :-------------------------- | :-------------------------------------------------------- |
| user.authentication.sso     | A user attempts a Single Sign-On (SSO) to an app managed in Okta |
| user.lifecycle.activate     | A user account is activated                             |
| user.lifecycle.create       | A user account is created                               |
| user.lifecycle.deactivate   | A user account is deactivated                           |
| user.lifecycle.suspend      | A user account is suspended                             |
| user.lifecycle.unsuspend    | A user account is moved from suspended status           |
| user.session.start          | Okta issues a session to a user who is authenticating    |

## Event correlation

When reviewing the System Log, it's often useful to correlate events to understand the thread of events that have occurred at a particular time.

The response object offers the following identifiers in this respect:

* `authenticationContext.externalSessionId`: Identifies events that occurred in the same user session

> * `transaction.id`: Identifies events that have occurred together as part of an operation (for example, a request to Okta's servers)

Reviewer: ```suggestion
* `transaction.id`: Identifies events that have occurred together as part of an operation (for example, a request to the Okta servers)
```

> * `authenticationContext.rootSessionId`: Identifies events from all sessions that share a common root session. Some actions that can be attributed to a user, such as those taken by an Okta system actor on behalf of a user may have a different `externalSessionId` value. The `rootSessionId` field allows all of those to be identified and grouped together.

Reviewer: ```suggestion
* `authenticationContext.rootSessionId`: Identifies events from all sessions that share a common root session. Some actions that can be attributed to a user, such as those taken by an Okta system actor on behalf of a user, may have a different `externalSessionId` value. The `rootSessionId` field allows all of those to be identified and grouped together.
```

| :-------------------------- | :-------------------------------------------------------- |
| user.authentication.sso     | A user attempts a Single Sign-On (SSO) to an app managed in Okta |
| user.lifecycle.activate     | A user account is activated                             |
| user.lifecycle.create       | A user account is created                               |
| user.lifecycle.deactivate   | A user account is deactivated                           |
| user.lifecycle.suspend      | A user account is suspended                             |
| user.lifecycle.unsuspend    | A user account is moved from suspended status           |
| user.session.start          | Okta issues a session to a user who is authenticating    |

#### User event details

`user.authentication.sso` doesn't capture whether the SSO attempt is successful or has failed, because Okta can't collect the subsequent authentication attempt status from the third-party service.

## Event correlation

When reviewing the System Log, it's often useful to correlate events to understand the thread of events that have occurred at a particular time.

The response object offers two identifiers in this respect:

>   - `authenticationContext.externalSessionId`: Identifies events that occurred in the same user session

Reviewer 2: Let's change this from two to three identifiers and add 

`authenticationContext.rootSessionId`: Identifies events from all sessions that share a common root session. Some actions that can be attributed to a user, such as those taken by an Okta system actor on behalf of a user may have a different `externalSessionId` value. The `rootSessionId` field allows all of those to be identified and grouped together.

### Correlating events based on API Token

It may be useful to identify multiple events that are the result of an action made using a specific API token. For example, when investigating a [rate limit warning](/docs/reference/rate-limits/), the events made by a specific token may be helpful in identifying the cause of the warning. The filter `filter=transaction.detail.requestApiTokenId eq "00T94e3cn9kSEO3c51s5"` returns all events that were the result of an action made using the token `00T94e3cn9kSEO3c51s5`, subject to other parameters of the query.

### Event correlation example

The following table shows 18 events produced from 13 transactions over six different sessions, all performed by one user.

**Note:** `authenticationContext.externalSessionId` is abbreviated to `sessionId` in this table.

| sessionId                   | transaction.id                | uuid                                   | Event Type                                     | displayMessage                        |
| :-------------------------- | :---------------------------- | :------------------------------------- | :-------------------------------------------- | :--------------------------------     |

> | trs5JnlvlaIQTOqOj9imLy7lA   | WcKPxq1f8QLfFvv3UPHhhgAACGM   | f24790d0-d324-47f8-aac5-c27a31ab928d   | user.session.access_admin_app                 | User accessing Okta administrator app |

Reviewer: ```suggestion
| trs5JnlvlaIQTOqOj9imLy7lA   | WcKPxq1f8QLfFvv3UPHhhgAACGM   | f24790d0-d324-47f8-aac5-c27a31ab928d   | user.session.access_admin_app                 | User accessing Okta Admin Console app |
```

> |                             | WcKPxq1f8QLfFvv3UPHhhgAACGM   | ed317758-8776-4240-a540-277c44dcb408   | application.lifecycle.update                  | Update application                    |

Reviewer: ```suggestion
|                             | WcKPxq1f8QLfFvv3UPHhhgAACGM   | ed317758-8776-4240-a540-277c44dcb408   | application.lifecycle.update                  | Update app                    |
```

> |                             |                               | 421c1551-71b2-4ebe-a70d-b5f7d3698429   | application.lifecycle.update                  | Update application                    |

Reviewer: ```suggestion
|                             |                               | 421c1551-71b2-4ebe-a70d-b5f7d3698429   | application.lifecycle.update                  | Update app                    |
```

> |                             |                               | 06a50bbe-44fd-40de-83e6-2e4cc2a17d16   | application.lifecycle.update                  | Update application                    |

Reviewer: ```suggestion
|                             |                               | 06a50bbe-44fd-40de-83e6-2e4cc2a17d16   | application.lifecycle.update                  | Update app                    |
```

> | trsUz2TG3wKS6ar1lvWzHo71w   | Wij-6q4YuniRd9yTmWHpfwAAADc   | 3e240ff4-6af7-47f2-b107-a2ef661ffc01   | application.user_membership.change_username   | Change users application username     |

Reviewer: ```suggestion
| trsUz2TG3wKS6ar1lvWzHo71w   | Wij-6q4YuniRd9yTmWHpfwAAADc   | 3e240ff4-6af7-47f2-b107-a2ef661ffc01   | application.user_membership.change_username   | Change user's app username     |
```

> |                             |                               | 572b05e9-b6be-4dfe-8bc3-01bb3a5a1af5   | application.user_membership.add               | Add user to application membership    |

Reviewer: ```suggestion
|                             |                               | 572b05e9-b6be-4dfe-8bc3-01bb3a5a1af5   | application.user_membership.add               | Add user to app membership    |
```

> |                             |                               | 30f29bbf-3218-429b-827a-0a93809591db   | application.user_membership.remove            | Remove users application membership   |

Reviewer: ```suggestion
|                             |                               | 30f29bbf-3218-429b-827a-0a93809591db   | application.user_membership.remove            | Remove users app membership   |
```

> |                             | Wij-964YuniRd9yTmWHu1AAAAEA   | 5f98d062-05a9-4ede-89a0-8a2ce27efdd4   | user.session.access_admin_app                 | User accessing Okta administrator app |

Reviewer: ```suggestion
|                             | Wij-964YuniRd9yTmWHu1AAAAEA   | 5f98d062-05a9-4ede-89a0-8a2ce27efdd4   | user.session.access_admin_app                 | User accessing Okta Admin Console app |
```

> |                             | Wij-95eCbHF7In2MKNavlgAAD9I   | 45f71ac2-e8b2-4c19-b4cc-d2560108c889   | application.lifecycle.update                  | Update application                    |

Reviewer: ```suggestion
|                             | Wij-95eCbHF7In2MKNavlgAAD9I   | 45f71ac2-e8b2-4c19-b4cc-d2560108c889   | application.lifecycle.update                  | Update app                    |
```

> |                             |                               | 46b85d65-01c6-44d2-86d2-25704804b1c5   | application.lifecycle.update                  | Update application                    |

Reviewer: ```suggestion
|                             |                               | 46b85d65-01c6-44d2-86d2-25704804b1c5   | application.lifecycle.update                  | Update app                    |
```

> | 102GALFw8CzRT2KXoqnca8Jdg   | Wij-AJeCbHF7In2MKNaOpAAAEC4   | b9ab9263-a4ae-4780-9981-377ec8f2da86   | user.session.start                            | User login to Okta                    |

Reviewer: ```suggestion
| 102GALFw8CzRT2KXoqnca8Jdg   | Wij-AJeCbHF7In2MKNaOpAAAEC4   | b9ab9263-a4ae-4780-9981-377ec8f2da86   | user.session.start                            | User signs in to Okta                    |
```

> |                             | Wij-7q4YuniRd9yTmWHrBQAAAKQ   | ff325685-0220-484c-82cf-5f8dc596acbe   | user.authentication.sso                       | User single sign on to the app            |

Reviewer: ```suggestion
|                             | Wij-7q4YuniRd9yTmWHrBQAAAKQ   | ff325685-0220-484c-82cf-5f8dc596acbe   | user.authentication.sso                       | User SSOs to the app            |
```

| trsf8nlpDJZTZeFlcc8nszbjw   | Wij-7a4YuniRd9yTmWHqqAAAAKY   | 5526a4c4-7f68-4b2a-bab7-2d10ebaeeb1c   | mim.checkOSXAccessEligibility.true            | *blank*                               |

> |                             | Wij-764YuniRd9yTmWHrkAAAAGw   | 232774ba-8feb-4b00-a732-e0ec99a24434   | user.session.start                            | User login to Okta                    |

Reviewer: ```suggestion
|                             | Wij-764YuniRd9yTmWHrkAAAAGw   | 232774ba-8feb-4b00-a732-e0ec99a24434   | user.session.start                            | User signs in to Okta                    |
```

> | trswPONv4wIRaKDNWVVcmtceg   | Wij-6K4YuniRd9yTmWHo9wAAAAY   | d31d819a-1427-45b0-a8b4-8a8fb40c72f1   | user.session.start                            | User login to Okta                    |

Reviewer: ```suggestion
| trswPONv4wIRaKDNWVVcmtceg   | Wij-6K4YuniRd9yTmWHo9wAAAAY   | d31d819a-1427-45b0-a8b4-8a8fb40c72f1   | user.session.start                            | User sign in to Okta                    |
```

|                             | Wij-564YuniRd9yTmWHoaQAAAII   | 0cc6f4c8-9b91-4a70-b5c4-09d6ad159d32   | mim.checkOSXAccessEligibility.true            | *blank*                               |

> |                             | Wij-2q4YuniRd9yTmWHjRAAAADA   | 92606da8-7eeb-4ad7-8ffb-502dd0ec64cc   | user.authentication.sso                       | User single sign on to the app            |

Reviewer: ```suggestion
|                             | Wij-2q4YuniRd9yTmWHjRAAAADA   | 92606da8-7eeb-4ad7-8ffb-502dd0ec64cc   | user.authentication.sso                       | User SSOs to the app            |
```

> | *null*                      | Wm@-R2s5lEMbNIB03krtvAAACyo   | 566671be-ec0b-400d-ad2e-6fc73ed12fb1   | user.session.start                            | User login to Okta                    |

Reviewer: ```suggestion
| *null*                      | Wm@-R2s5lEMbNIB03krtvAAACyo   | 566671be-ec0b-400d-ad2e-6fc73ed12fb1   | user.session.start                            | User signs in to Okta                    |
```

## List events

Use the [List all System Log Events](https://developer.okta.com/docs/api/openapi/okta-management/management/tag/SystemLog/#tag/SystemLog/operation/listLogEvents) endpoint to retrieve a list of events from your org. By default, this request returns a limit of 100 response objects.

##### Request types

All requests to the `/api/v1/logs` endpoint fall into one of these two categories:

* [Polling Requests](#polling-requests)
* [Bounded Requests](#bounded-requests)

###### Polling requests

Polling requests are for situations when you want to consume an ongoing stream of events from Okta.

Example use cases include:

* [Ingesting System Log data into an external SIEM system](#transfer-data-to-a-separate-system).
* Using System Log data for real-time monitoring.

For a request to be a polling request, it must meet the following request parameter criteria:

> * `until` must be unspecified.

Reviewer: ```suggestion
* `until` is unspecified.
```

* `sortOrder` is `ASCENDING`.

Polling requests to the `/api/v1/logs` API have the following semantics:

* They return every event that occurs in your org.

> * The returned events are time filtered by their internal "persistence time" to avoid skipping records due to system delays (unlike [Bounded Requests](#bounded-requests)).

Reviewer 2: Update for clarity:
The returned events are ordered by the internal "persistence time" which reflects when an event is actually committed to the log, and not necessarily the timestamp when the event occurred. The persistence time may differ from the externally available `published` timestamp.

* They may return events out of order according to the `published` field.
* They have an infinite number of pages. That is, a [`next` `link` relation header](#next-link-response-header) is always present, even if there are no new events (the event list may be empty).

###### Bounded requests

Bounded requests are for situations when you know the definite time period of logs you want to retrieve.

Example use cases include:

* [Debugging or troubleshooting system behavior](#debugging)
* Auditing events that happened at a particular time

For a request to be a bounded request, it must meet the following request parameter criteria:

* `since` must be specified

> * `until` must be specified

Reviewer: ```suggestion
* `until` is specified
```

Bounded requests to the `/api/v1/logs` API have the following semantics:

> * The returned events are time filtered by their associated `published` field (unlike [Polling Requests](#polling-requests)).

Reviewer 2: Updated to say "ordered and filtered"

* The returned events are guaranteed to be in order according to the `published` field.
* They have a finite number of pages. That is, the last page doesn't contain a [`next` `link` relation header](#next-link-response-header).
* Not all events for the specified time range may be present. Some events may be delayed. Such delays are rare but possible.

##### Filter results

###### Expression filter

> An expression filter is useful for performing structured queries where constraints on the response object's attribute values can be explicitly targeted.

Reviewer: ```suggestion
An expression filter is useful for performing structured queries where you can explicitly target constraints on the response object's attribute values.
```

> **Note:** SCIM filter expressions can't use the `published` attribute since it may conflict with the logic of the `since`, `after`, and `until` query parameters.
> In addition, a SCIM filter expression that uses the `co` (contains) operator with the `debugContext.debugData.url` or the `debugContext.debugData.requestUri` attribute isn’t supported.
> A request with an invalid SCIM filter expression returns an HTTP 400 API response.

See [Filtering](/docs/reference/core-okta-api/#filter) for more information on expressions.

The following are examples of filter expressions:

* Events that are published for a target user

```javascript
filter=target.id eq "00uxc78lMKUMVIHLTAXY"
```

* Events that are published for all actors except for a specific user

```javascript
filter=actor.id ne "00uxc78lMKUMVIHLTAXY"
```

* Failed sign-in events

```javascript
filter=eventType eq "user.session.start" and outcome.result eq "FAILURE"
```

> * Events that are published for a target user and application

Reviewer: ```suggestion
* Events that are published for a target user and app
```

```javascript
filter=target.id eq "00uxc78lMKUMVIHLTAXY" and target.id eq "0oabe82gnXOFVCDUMVAK"
```

* App SSO events for a target user and app

```javascript

> filter=eventType eq "app.auth.sso" and target.id eq "00uxc78lMKUMVIHLTAXY" and target.id eq "0oabe82gnXOFVCDUMVAK"

Reviewer 2: update to `user.authentication.sso` as `app.auth.sso` is not a valid event type

> * App SSO events for a target user and application

Reviewer: ```suggestion
* App SSO events for a target user and app
```

```

* Events that are published for a given IP address

```javascript
filter=client.ipAddress eq "184.73.186.14"
```

* Events that start with `event_hook`

```javascript
filter=eventType sw "event_hook"
```

* Events that contain session

```javascript
filter=eventType co "session"
```

* Events that end with a token

```javascript
filter=eventType ew "token"
```

###### Keyword filter

> The query parameter `q` can be used to perform keyword matching against a response object's attribute values. All supplied keywords must be matched exactly.

Reviewer: ```suggestion
You can use the query parameter `q` to perform keyword matching against a response object's attribute values. All supplied keywords must be matched exactly.
```

>**Note:** Keyword matching is case-insensitive.

The following are examples of common keyword filtering:

* Events that mention a specific city: `q=San Francisco`
* Events that mention a specific URL: `q=interestingURI.com`
* Events that mention a specific person: `q=firstName lastName`

> > **Note:** When hyphens are present in an event instance's attribute value, they’re split and added to the list of matching candidates, in addition to the full hyphenated value. Therefore, events that contain the text `XOxBw-2JIRnCFd0gG0GjHAAABjY` are matched with a `q` value of `XOxBw`, `2JIRnCFd0gG0GjHAAABjY`, or `XOxBw-2JIRnCFd0gG0GjHAAABjY`.

Reviewer: ```suggestion
> **Note:** When hyphens are present in an event instance's attribute value, they’re split and added to the list of matching candidates, in addition to the fully hyphenated value. Therefore, events that contain the text `XOxBw-2JIRnCFd0gG0GjHAAABjY` are matched with a `q` value of `XOxBw`, `2JIRnCFd0gG0GjHAAABjY`, or `XOxBw-2JIRnCFd0gG0GjHAAABjY`.
```

###### Date and time filter

> The System Log response objects can be filtered by the response object's [published](https://developer.okta.com/docs/api/openapi/okta-management/management/tag/SystemLog/#tag/SystemLog/operation/listLogEvents!c=200&path=published&t=response) attribute value with the following combination of parameters:

Reviewer: ```suggestion
You can filter the System Log response objects by the response object's [published](https://developer.okta.com/docs/api/openapi/okta-management/management/tag/SystemLog/#tag/SystemLog/operation/listLogEvents!c=200&path=published&t=response) attribute value with the following combination of parameters:
```

}
```

An invalid SCIM field and operator combination within a `filter` request parameter (for example, `debugContext.debugData.url co "/oauth/"`) returns an HTTP 400 error with a message that indicates the unsupported combination, for example:

```json
{
  "errorCode": "E0000031",
  "errorSummary": "The supplied combination of operator and field is not currently supported. Operator: co, Field: debugContext.debugData.url",
  "errorId": "ec93dhe2-6d76-458c-8c0c-f8df8fb7a24b"
}
```

An internal service error returns the HTTP 500 error with the message:

```json
{
  "errorCode": "E0000053",
  "errorSummary": "Sorry, there's been an error. We aren't sure what caused it, but we've logged this and will work to address it. Please try your request again.",
  "errorId": "55166534-b7d8-45a5-a4f6-3b38a5507046"
}
```

A timeout returns the HTTP 500 error with the message:

```json
{
  "errorCode": "E0000009",
  "errorSummary": "Your last request took too long to complete. This is likely due to a load issue on our side. We've logged this and will work to address it. Please either simplify your query or wait a few minutes and try again."
}
```

> A free-form query that is too long returns the following error message:

Reviewer: ```suggestion
A free-form query that's too long returns the following error message:
```

```json
{
  "errorCode": "E0000001",
  "errorSummary": "Api validation failed: 'q': Freeform search cannot contain items longer than 40 characters. Please shorten the items in your search or use an advanced filter to query by specific fields."
}
```

Exceeding the rate limit returns the following error message:

```json
{
  "errorCode": "E0000047",
  "errorSummary": "API call exceeded rate limit due to too many requests."
}
```

## Data retention

System Log data older than 90 days isn't returned, in accordance with Okta's [Data Retention Policy](https://support.okta.com/help/Documentation/Knowledge_Article/Okta-Data-Retention-Policy). Queries that exceed the retention period succeed, but only those results that have a `published` timestamp within the window are returned.

## Examples

### Debugging

> The System Log API can be used to troubleshoot user problems. For example, you can use the following `curl` command to see events from user "Jane Doe":

Reviewer: ```suggestion
Use the System Log API to troubleshoot user problems. For example, you can use the following `curl` command to see events from user "Jane Doe":
```
